src/contracts/SavingCirclesViewer.sol

```diff
@@ -0,0 +1,349 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {SavingCircles} from 'contracts/SavingCircles.sol';
+
+import {ISavingCircles} from 'interfaces/ISavingCircles.sol';
+import {ISavingCirclesViewer} from 'interfaces/ISavingCirclesViewer.sol';
+
+/**
+ * @title Saving Circles Viewer
+ * @notice Contract for viewing the state of the Saving Circles contract
+ * @dev This contract is used exclusively to view the state of the Saving Circles contract
+ * @dev This contract does not modify or interact with the Saving Circles contract
+ * @author Breadchain Collective
+ * @author @RonTuretzky
+ * @author bagelface.eth
+ * @author exo404
+ * @author valeriooconte
+ */
+contract SavingCirclesViewer is ISavingCirclesViewer {
+  SavingCircles public immutable SAVING_CIRCLES;
+
+  constructor(address _savingCircles) {
+    SAVING_CIRCLES = SavingCircles(_savingCircles);
+  }
+
+  /// @inheritdoc ISavingCirclesViewer
+  function isTokenAllowed(address _token) external view override returns (bool) {
+    return SAVING_CIRCLES.allowedTokens(_token);
+  }
+
+  /// @inheritdoc ISavingCirclesViewer
+  function checkMemberships(
+    address _member,
+    uint256[] calldata _ids
+  ) external view override returns (bool[] memory _statuses) {
+    _statuses = new bool[](_ids.length);
+
+    for (uint256 i = 0; i < _ids.length; i++) {
+      _statuses[i] = SAVING_CIRCLES.isMember(_ids[i], _member);
+    }
+
+    return _statuses;
+  }
+
+  /// @inheritdoc ISavingCirclesViewer
+  function getComprehensiveUserData(address _user)
+    external
+    view
+    override
+    returns (ComprehensiveUserData memory userData)
+  {
+    userData.userAddress = _user;
+    userData.timestamp = block.timestamp;
+    userData.blockNumber = block.number;
+
+    uint256[] memory userCircleIds = _getAllUserCircleIds(_user);
+    userData.circleData = new UserCircleData[](userCircleIds.length);
+
+    userData.membershipStatus = _getUserMembershipStatus(_user, userCircleIds);
+    userData.financialSummary = _getUserFinancialSummary(_user, userCircleIds);
+
+    for (uint256 i = 0; i < userCircleIds.length; i++) {
+      userData.circleData[i] = _getUserCircleData(_user, userCircleIds[i]);
+    }
+
+    return userData;
+  }
+
+  /// @inheritdoc ISavingCirclesViewer
+  function getUserFinancialSummary(address _user) external view override returns (UserFinancialSummary memory summary) {
+    uint256[] memory userCircleIds = _getAllUserCircleIds(_user);
+    return _getUserFinancialSummary(_user, userCircleIds);
+  }
+
+  /// @inheritdoc ISavingCirclesViewer
+  function getUserMembershipStatus(address _user) external view override returns (UserMembershipStatus memory status) {
+    uint256[] memory userCircleIds = _getAllUserCircleIds(_user);
+    return _getUserMembershipStatus(_user, userCircleIds);
+  }
```

Comment:
we might want to delete these, pretty sure the frontend is going to just use the`getComprehensiveUserData` function